### PR TITLE
feat(api): implement RedisBackup CRD with S3 backup controller

### PR DESCRIPTION
**Pull request by Rohit-Kiran24**
## Summary
Implements the `RedisBackup` operator — a new Kubernetes controller that allows
users to define and trigger backups of Redis clusters to S3-compatible storage.

## New API
```yaml
apiVersion: redis.redis.opstreelabs.in/v1alpha1
kind: RedisBackup
metadata:
  name: redis-backup-sample
spec:
  redisClusterName: redis-cluster
  storageType: s3
  s3:
    bucket: my-redis-backups
    region: ap-south-1
    secretName: aws-s3-credentials
  retentionDays: 7
```

## Changes
- `api/v1alpha1/redisbackup_types.go` — Full CRD spec with S3 config, schedule, retention, status fields
- `internal/controller/redisbackup_controller.go` — Reconcile loop with validation, phase transitions, idempotency, and S3 routing
- `internal/controller/suite_test.go` — envtest suite with manager and reconciler setup
- `internal/controller/redisbackup_controller_test.go` — Tests for success, missing S3 config, and API validation rejection
- `cmd/main.go` — Registered RedisBackupReconciler with the controller manager
- Generated: CRD manifest, RBAC policies, sample YAML

## Test results
All controller tests pass, including envtest-based integration tests.

### CODE REVIEW

**Comment by rmalisetti** [comment]
Hey @Rohit-Kiran24 , Hope you are doing well!
I'm looking for a backup solution from the Operator and came across your PR. Are you looking for something to get this PR merged? And any tentative timelines when this can be part of operator? Let me know!

@shubham-cmyk  @drivebyer  @iamabhishek-dubey - Can we get some initial review on this PR. Thanks!

Regards,
Ram

**Comment by Rohit-Kiran24** [comment]
Hey @rmalisetti, thanks for the interest and for pinging the reviewers — really appreciate it! 🙏

The PR currently has the CRD scaffolding, API types, and controller structure in place. I'm actively working on a round of improvements before requesting final review — specifically around controller registration, validation, structural alignment with existing patterns, and RBAC polish.

I'll push the updates shortly. No hard ETA for merge since it depends on maintainer review cycles, but I'm targeting to have the updated PR ready for review within this week.

Happy to collaborate if you have specific backup requirements or use cases in mind!

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 can you rebase the branch with main

**Comment by Rohit-Kiran24** [comment]
Sure @shubham-cmyk, will rebase with main and fix the CI failures. Give me a moment!

**Comment by Rohit-Kiran24** [comment]
> @Rohit-Kiran24 can you rebase the branch with main

Done @shubham-cmyk! Rebased with main, fixed lint issues, and added RBAC markers for controller-gen. Ready for review.

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 This is going to be pretty big review. i am doing in part's i would put the comments one by one.

File: api/redisbackup/v1alpha1/groupversion_info.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*
```
Comment (shubham-cmyk) [inline api/redisbackup/v1alpha1/groupversion_info.go:1]: Move this file into the 
`api/redisbackup/v1alpha1`

File: internal/controller/redisbackup/redisbackup_controller.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+// backupToS3 resolves the S3 backup location.
+//
+// TODO(alpha): This currently only constructs the expected S3 path and
+// validates that credentials exist. The full implementation should:
+//  1. Trigger BGSAVE on the target Redis instance
+//  2. Wait for the RDB snapshot to complete
+//  3. Copy the RDB file from the Redis pod
+//  4. Upload it to S3 using the AWS SDK with credentials from the referenced Secret
+//  5. Verify the upload integrity (e.g., checksum)
+func (r *Reconciler) backupToS3(ctx context.Context, instance *redisv1alpha1.RedisBackup) (string, error) {
```
Comment (shubham-cmyk) [inline internal/controller/redisbackup/redisbackup_controller.go]: Missing BGSAVE cmd to run

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 check lint failures, run make codegen

File: internal/controller/redisbackup/redisbackup_controller.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+		Message:            reason,
+	})
+
+	return r.Status().Update(ctx, instance)
+}
+
+// SetupWithManager registers this controller with the controller manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, opts controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&redisv1alpha1.RedisBackup{}).
```
Comment (shubham-cmyk) [inline internal/controller/redisbackup/redisbackup_controller.go]: `SetupWithManager` only does `For(&RedisBackup{})`, and a `validateSpec` failure returns `ctrl.Result{}, nil` (no requeue). So a backup created before its Secret exists goes `Failed` and stays `Failed` forever, even after the Secret is created — until the CR is edited. Consider `Watches`/`Owns` on the referenced Secret.

File: api/v1alpha1/redisbackup_types.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// StorageType defines the backend storage provider for backups
+// +kubebuilder:validation:Enum=s3;gcs;azure
```
Comment (shubham-cmyk) [inline api/v1alpha1/redisbackup_types.go]: don't include the gcs and azure here right now if it is not implemented.

File: internal/controller/redisbackup/redisbackup_controller.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+
+// Reconciler reconciles a RedisBackup object.
+type Reconciler struct {
+	client.Client
+	K8sClient kubernetes.Interface
+}
+
+// RBAC permissions required by this controller.
+//+kubebuilder:rbac:groups=redis.redis.opstreelabs.in,resources=redisbackups,verbs=get;list;watch;create;update;patch;delete
```
Comment (shubham-cmyk) [inline internal/controller/redisbackup/redisbackup_controller.go]: these are duplicate we already have rbac at L 45-48

**Review by shubham-cmyk — CHANGES_REQUESTED** [review]

File: api/redisbackup/v1alpha1/groupversion_info.go (hunk 1 above)
Comment (Rohit-Kiran24) [reply api/redisbackup/v1alpha1/groupversion_info.go:1]: Done, moved all API types to api/redisbackup/v1alpha1/ to match existing patterns (api/redis/v1beta2/, etc.).

File: internal/controller/redisbackup/redisbackup_controller.go (hunk 2 above)
Comment (Rohit-Kiran24) [reply internal/controller/redisbackup/redisbackup_controller.go]: Added BGSAVE execution via pods/exec on the Redis pod. Full exec + S3 upload flow is marked as TODO for a follow-up.

File: internal/controller/redisbackup/redisbackup_controller.go (hunk 3 above)
Comment (Rohit-Kiran24) [reply internal/controller/redisbackup/redisbackup_controller.go]: Good catch! Added Watches(&corev1.Secret{}) in SetupWithManager — the controller now re-reconciles when the referenced Secret is created/updated. Also changed validation failure to requeue after 30s instead of returning nil.

File: api/v1alpha1/redisbackup_types.go (hunk 4 above)
Comment (Rohit-Kiran24) [reply api/v1alpha1/redisbackup_types.go]: Removed — only s3 in the enum now. Will add gcs/azure when implemented.

File: internal/controller/redisbackup/redisbackup_controller.go (hunk 5 above)
Comment (Rohit-Kiran24) [reply internal/controller/redisbackup/redisbackup_controller.go]: Removed the duplicates. Added pods/exec RBAC marker for BGSAVE.

**Comment by Rohit-Kiran24** [comment]
@shubham-cmyk All review feedback addressed. Fixed lint (0 issues), ran codegen, and rebased. Ready for re-review 🙏

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 the `golang ci lint`  and `codegen` is still failing, otherwise looks fine. 

What is plan to add the BGSAVE cmd here ?

**Comment by Rohit-Kiran24** [comment]
Hi @shubham-cmyk,

CI Fixes: Just pushed a fix — verify-codegen was failing because I hadn't run the full make codegen pipeline (sync-crds, generate-dataAssert, generate-metricsdocs, generate-api-docs). All codegen outputs are now included. Semantic PR validation and lint are also addressed.

BGSAVE Plan: Here's the plan for the full BGSAVE implementation:

1)Execute redis-cli BGSAVE on the target Redis pod via pods/exec
2)Poll LASTSAVE to confirm the RDB snapshot completed
3)Copy the dump.rdb file from the pod using tar stream
4)Upload the RDB file to S3 using the AWS SDK with credentials from the referenced Secret
5)Verify upload integrity (ETag/checksum)
The scaffolding for step 1 is already in place with the required pods/exec RBAC permission.

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 the plan seems fine to be there are few tools like restic which i used earlier for backup you can look at that, meanwhile how do you plan to restore the backup. this has been a tricky part.

**Comment by Rohit-Kiran24** [comment]
@shubham-cmyk, great suggestion — I'll look into restic for the backup layer. It would simplify multi-backend support (S3, GCS, Azure) and add deduplication + encryption out of the box.

Restore Plan:

Agreed, restore is the tricky part. Here's my approach:

1. Separate RedisRestore CRD — keeps backup and restore as independent concerns (similar to Velero's Backup/Restore pattern).

2. Restore flow:

User creates a RedisRestore CR pointing to a backupLocation (e.g. s3://bucket/path/dump.rdb)
Controller scales down the target Redis StatefulSet to 0 replicas
Downloads the RDB from S3 into the Redis PVC data directory as dump.rdb
Scales Redis back up — Redis natively loads the RDB on startup
Controller polls Redis health, verifies data is loaded, then marks RestoreCompleted

3. Edge cases:

Cluster mode: restore to the leader pod first, let replication sync followers
Sentinel mode: pause failover during restore to avoid split-brain
Rollback safety: snapshot existing data before overwriting so we can roll back if restore fails

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 AFAIK we have to restore the `node.conf` also there, can you try this and raise the PR i would test end to end on my side also

**Comment by Rohit-Kiran24** [comment]
@shubham-cmyk, good catch — node.conf backup and restore is now included. Here's how it works:

What's Been Added
Backup: BGSAVE → poll LASTSAVE → copy dump.rdb + node.conf from pod → upload to S3

Restore: Download from S3 → scale down StatefulSet → restore dump.rdb + node.conf to PVC → scale back up → verify health

Compatibility: node.conf handling is conditional — works seamlessly for both standalone and cluster mode

 Cluster Mode Note: Currently node.conf is restored as-is, which works for same-topology restores. For cross-cluster restores where Pod IPs change, dynamic patching of node.conf (updating IPs/hostnames) would be needed — I can add that in a follow-up if required.

How to Test End-to-End
1. Create S3 credentials Secret:
apiVersion: v1
kind: Secret
metadata:
  name: redis-s3-credentials
type: Opaque
stringData:
  AWS_ACCESS_KEY_ID: "<your-key>"
  AWS_SECRET_ACCESS_KEY: "<your-secret>"

2. Trigger backup:
kubectl apply -f config/samples/redis_v1alpha1_redisbackup.yaml
kubectl get redisbackup -w  # watch until Phase=Completed

3. Trigger restore:
kubectl apply -f config/samples/redis_v1alpha1_redisrestore.yaml
kubectl get redisrestore -w  # watch until Phase=Completed

The s3.endpoint field supports MinIO for local testing. Let me know how the E2E goes on your side!

**Comment by shubham-cmyk** [comment]
> Cluster Mode Note: Currently node.conf is restored as-is, which works for same-topology restores. For cross-cluster restores where Pod IPs change, dynamic patching of node.conf (updating IPs/hostnames) would be needed — I can add that in a follow-up if required.

yeah I wanted to see this how does it go can you complete the PR

**Comment by Rohit-Kiran24** [comment]
@shubham-cmyk Sure, will implement the dynamic node.conf patching for cross-cluster restores (updating IPs/hostnames to match current pod topology). Give me some time to get this done properly — will push the update and let you know!

**Comment by Rohit-Kiran24** [comment]
@shubham-cmyk , I've wrapped up the implementation for both flows! Here is a detailed breakdown of how the architecture is structured and exactly how you can test it end-to-end.

##  Architecture & Component Breakdown

We're introducing two new CRDs to keep the backup and recovery concerns cleanly separated:

- RedisBackup (short name: rb) — Backs up Redis data + cluster state to S3

- RedisRestore (short name: rr) — Restores Redis from an existing S3 backup package

---

###  The Backup Flow (`RedisBackup`)

When a `RedisBackup` CR is created, the controller executes the following lifecycle:

1. Validation:---Checks the storage type, S3 configuration, and ensures the referenced secret contains valid `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` fields.

2. Snapshot Creation:-- Executes `redis-cli BGSAVE` on the `<clusterName>-0` pod via a custom `pods/exec` SPDY stream.

3. Completion Polling:--Polls `LASTSAVE` every 2 seconds (with a 60-second timeout) until the unix timestamp increments, confirming the RDB file is fully written.

4. File Streaming:--Runs a `tar cf` command via execution context to stream `dump.rdb` and `node.conf` (if running in cluster mode) from `/data/` down to a local temporary directory.

5. Cloud Upload:-- Uses the AWS SDK v2 to push both files securely to `s3://<bucket>/backups/<clusterName>/<timestamp>/`.

6. Status Update:--Marks the resource as `Phase: Completed` and exposes the exact path in `status.backupLocation`.

---

###  The Restore Flow (`RedisRestore`)

The recovery pipeline handles the trickier orchestration required to avoid data corruption or split-brain states:

1. Validation:-- Verifies the connection to the target backup location and cross-checks the AWS storage credentials.

2. --Download:--Stages the remote `dump.rdb` and `node.conf` locally within the controller's runtime environment.

3. Scale-Down Safety:--Scales the target Redis StatefulSet down to `0` replicas and blocks until all pods gracefully terminate, completely releasing the PVC locks.

4. Target Isolation:--Re-scales the StatefulSet specifically to `1` replica to get `pod-0` bound to its PVC, then immediately hits it with a `redis-cli SHUTDOWN NOSAVE` to drop the active process before it can write any fresh data.

5. Data Injection:--Streams the backup bundle into the pod's `/data/` directory using an automated `tar xf` stream.

6.  Dynamic Topology Patching:--Reads the restored `node.conf` from the pod, uses regex to parse out all stale `<ip>:<port>@<bus-port>` entries, replaces them inline with the current `pod.Status.PodIP`, and writes it back. This prevents cluster linkage failures during cross-topology or fresh-infra restores.

7. Re-scale & Validation:-- Scales the StatefulSet back up to its original replica count, waits for readiness, and flags the CR as `Phase: Completed`.

---

## How to Test End-to-End

#### Step 1: Create your S3 credentials Secret

Save the following as `secret.yaml` and apply it:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: redis-s3-credentials
  namespace: default
type: Opaque
stringData:
  AWS_ACCESS_KEY_ID: "<your-key>"
  AWS_SECRET_ACCESS_KEY: "<your-secret>"
```

```bash
kubectl apply -f secret.yaml
```

#### Step 2: Trigger and watch a Backup

Save the following as `backup.yaml` (adjust fields as needed):

```yaml
apiVersion: redis.redis.opstreelabs.in/v1alpha1
kind: RedisBackup
metadata:
  name: my-backup
  namespace: default
spec:
  redisClusterName: <your-redis-cluster>
  storageType: s3
  s3:
    bucket: <your-bucket>
    region: <your-region>
    secretName: redis-s3-credentials
    # endpoint: "http://minio:9000"  # Optional: uncomment this if testing locally with MinIO
```

```bash
kubectl apply -f backup.yaml
kubectl get redisbackup my-backup -w

# Once PHASE shows "Completed", extract the exact backup path:
kubectl get redisbackup my-backup -o jsonpath='{.status.backupLocation}'
```

#### Step 3: Execute the Restore

Create your `restore.yaml` using the backup path obtained from the step above:

```yaml
apiVersion: redis.redis.opstreelabs.in/v1alpha1
kind: RedisRestore
metadata:
  name: my-restore
  namespace: default
spec:
  redisClusterName: <your-redis-cluster>
  storageType: s3
  backupLocation: "s3://<bucket>/backups/<cluster>/<timestamp>"
  s3:
    bucket: <your-bucket>
    region: <your-region>
    secretName: redis-s3-credentials
```

```bash
kubectl apply -f restore.yaml
kubectl get redisrestore my-restore -w
```

****#### Step 4: Verify Cluster Consistency****

Once the restore phase finishes, verify that data integrity is sound and cluster links are properly re-established:

```bash
kubectl exec -it <cluster>-0 -- redis-cli DBSIZE
kubectl exec -it <cluster>-0 -- redis-cli CLUSTER NODES  # If evaluating cluster topology
```

---

Let me know how the end-to-end testing goes on your cluster. I'm happy to refine any part of the logic based on your findings!

---

**Comment by rmalisetti** [comment]
@Rohit-Kiran24 @shubham-cmyk  - Just a thought! How about taking the backup of Redis external config as-well along with node conf? Let me know your thoughts, thank you!

-Ram

**Comment by Rohit-Kiran24** [comment]
@rmalisetti Good call! Just pushed an update — the backup now captures redis.conf (external Redis config) alongside dump.rdb and node.conf. All three are handled conditionally, so it works seamlessly for standalone, cluster, and custom-config setups. On restore, the config is placed back before Redis starts up. Thanks for the suggestion! 

@shubham-cmyk FYI — this update also includes a rebase on latest main (picked up the 5 recent commits). Ready for your E2E testing whenever you get a chance!

**Comment by shubham-cmyk** [comment]
sorry for delay had a busy last month, continuing on looking where i left on this

**Comment by Rohit-Kiran24** [comment]
Thanks @shubham-cmyk, no worries about the delay — and thanks for the rebase.

Heads-up: two files in my commit had been mangled by a Windows shell redirect.

        -  charts/redis-operator/crds/crds.yaml had every line break stripped — the whole 1.59 MB file on one line, so the YAML wouldn't parse and helm install would have failed.

       - docs/content/en/docs/Monitoring/metrics.md got re-encoded to UTF-16, which is why it showed as a binary diff.

Both fixed and pushed (8449911), rebased on latest main. The chart diff is now +405 / −0 — just the two new CRDs, nothing else touched. Codegen is clean and go build/go vet pass.

Two things on your side:

       - The 4 required workflows are awaiting approval — they need a maintainer to start them.

       - The changes-requested review is still blocking merge. I addressed that feedback back in June, so a re-review would unblock it.

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 are you opstree slack channel. you can say hi to me in the redis-operator group. i have to collaborate with you on this PR.

File: internal/backuputil/exec.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+	return nil
+}
+
+// extractTar writes a tar stream into destDir, rejecting any entry whose path
+// escapes destDir. Entry names come from inside the container, so they are
+// treated as untrusted input.
+func extractTar(r io.Reader, destDir string) error {
+	tr := tar.NewReader(r)
+	for {
+		hdr, err := tr.Next()
```
[the remark this answers is not in the document]
Comment (shubham-cmyk) [reply internal/backuputil/exec.go:328]: fixed

**Comment by Rohit-Kiran24** [comment]
> @shubham-cmyk No worries at all about the delay, welcome back!

> Yes, I am in the Opstree Slack workspace! However, when I search for "Shubham gupta", there are several people with that exact name. Could you let me know your exact Slack handle or profile picture so I can make sure I ping the right person?

> Thanks for jumping in and making those refactoring commits. I saw the Copilot review caught some great edge cases (the OOM streaming risk and the StatefulSet reconcile races). Let's sync up on Slack and divide up the remaining fixes and unit tests!

**Comment by shubham-cmyk** [comment]
@Rohit-Kiran24 I have texted you in opstree slack check
